src/linear_algebra/multilinear.lean

```diff
@@ -583,16 +583,24 @@ end distrib_mul_action
 section semimodule
 
 variables {R' A : Type*} [semiring R'] [semiring A]
-  [Π i, semimodule A (M₁ i)] [semimodule R' M₂] [semimodule A M₂] [smul_comm_class A R' M₂]
+  [Π i, semimodule A (M₁ i)] [semimodule A M₂]
+  [add_comm_monoid M₃] [semimodule R' M₃] [semimodule A M₃] [smul_comm_class A R' M₃]
 
 /-- The space of multilinear maps over an algebra over `R` is a module over `R`, for the pointwise
 addition and scalar multiplication. -/
-instance : semimodule R' (multilinear_map A M₁ M₂) :=
+instance [semimodule R' M₂] [smul_comm_class A R' M₂] : semimodule R' (multilinear_map A M₁ M₂) :=
 { add_smul := λ r₁ r₂ f, ext $ λ x, add_smul _ _ _,
   zero_smul := λ f, ext $ λ x, zero_smul _ _ }
 
-end semimodule
+/-- `multilinear_map.dom_dom_congr` as a `linear_equiv`. -/
+@[simps apply symm_apply]
+def dom_dom_congr_linear_equiv {ι₁ ι₂} [decidable_eq ι₁] [decidable_eq ι₂] (σ : ι₁ ≃ ι₂) :
+  multilinear_map A (λ i : ι₁, M₂) M₃ ≃ₗ[R'] multilinear_map A (λ i : ι₂, M₂) M₃ :=
```

Comment:
Oh, using `smul_comm_class` is a clever trick to get around issues with commutativity of `R`. Perhaps we should go back and do this for all the `add_equiv`s that we couldn't quite make `linear_equiv`s.

Should `R'` be an explicit argument?

Reply:
I made most type args explicit because Lean fails to infer them in the definition of `curry_fin_finset`.